src/microp.h

```diff
@@ -0,0 +1,193 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef MICROP_H_
+#define MICROP_H_
+
+#include "natsp.h"
+#include "mem.h"
+
+#define MICRO_CALL(__err, __call) \
+    if ((__err) == NULL)          \
+    {                             \
+        __err = (__call);         \
+    }
+
+#define MICRO_DO(__err, __block) \
+    if ((__err) == NULL)         \
+        __block;
+
+#define MICRO_QUEUE_GROUP "q"
+
+#define MICRO_DEFAULT_ENDPOINT_NAME "default"
+
+// 4 verbs: INFO, STATS, PING, SCHEMA;
+// 3 subjects for each verb.
+#define MICRO_MONITORING_SUBS_CAP (4 * 3)
+
+struct micro_error_s
+{
+    struct micro_error_s *cause;
+    natsStatus status;
+    int code;
+    char *message;
+};
+
+struct micro_client_s
+{
+    natsConnection *nc;
+};
+
+struct micro_endpoint_s
+{
+    // The name and subject that the endpoint is listening on (may be different
+    // from one specified in config).
+    char *name;
+    char *subject;
+
+    // References to other entities.
+    microService *m;
+    microEndpointConfig *config;
+
+    // Monitoring endpoints are different in a few ways. For now, express it as
+    // a single flag but consider unbundling:
+    //   - use_queue_group: Service endpoints use a queue group, monitoring
+    //     endpoints don't.
+    //   - forward_response_errors_to_async_handler: Service endpoints handle
+    //     respond errors themselves, standard monitoring endpoints don't, so
+    //     send the errors to the service async handler, if exists.
+    //   - gather_stats: Monitoring endpoints don't need stats.
+    //   - include_in_info: Monitoring endpoints are not listed in INFO
+    //     responses.
+    bool is_monitoring_endpoint;
+
+    // Mutex for starting/stopping the endpoint, and for updating the stats.
+    natsMutex *endpoint_mu;
+    int refs;
+
+    // The subscription for the endpoint. If NULL, the endpoint is stopped.
+    natsSubscription *sub;
+
+    // Endpoint stats. These are initialized only for running endpoints, and are
+    // cleared if the endpoint is stopped.
+    microEndpointStats stats;
+
+    microEndpoint *next;
+};
+
+struct micro_group_s
+{
+    struct micro_service_s *m;
+    struct micro_group_s *next;
+    char prefix[];
+};
+
+struct micro_service_s
+{
+    // these are set at initialization time time and do not change.
+    natsConnection *nc;
+    microServiceConfig *cfg;
+    char id[NUID_BUFFER_LEN + 1];
+
+    // groups are just convenient wrappers to make "prefixed" endpoints with
+    // AddEndpoint. They are added at initializaton time, so no need to lock.
+    struct micro_group_s *groups;
+
+    // these are are updated concurrently with access as the service runs, so
+    // need to be protected by mutex.
+    natsMutex *service_mu;
+    int refs;
+
+    struct micro_endpoint_s *first_ep;
+    int num_eps;
+
+    natsConnectionHandler prev_on_connection_closed;
```

Comment:
Comment about naming that I made earlier. Maybe `user_.._cb` would have been better?